module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -278,36 +277,47 @@ namespace app
         }
     }
 
-    void ApplicationCommon::returnToPreviousWindow(const uint32_t times)
+    /// wait we have it already in some class
+    /// BUT SHIT IT'S NOT PROPAGATED :(((
+    void ApplicationCommon::switchWindowPopup(const std::string &windowName,
+                                              const gui::popup::Disposition &d,
+                                              std::unique_ptr<gui::SwitchData> data,
+                                              SwitchReason reason)
+    {
+        bus.sendUnicast(std::make_shared<AppSwitchWindowPopupMessage>(windowName, std::move(data), reason, d),
+                        this->GetName());
+    }
+
+    void ApplicationCommon::returnToPreviousWindow()
     {
-        if (const auto prevWindow = getPrevWindow(times); prevWindow == gui::name::window::no_window) {
+        windowsStack().logStack(std::cout);
+        auto window = windowsStack().get(previousWindow);
+        if (not window) {
+            LOG_DEBUG("No window to back from - get to previous app");
             app::manager::Controller::switchBack(this);
+            return;
         }
-        else {
-            LOG_INFO("Back to previous window %s", prevWindow.c_str());
-            switchWindow(prevWindow, gui::ShowMode::GUI_SHOW_RETURN);
-        }
+        LOG_INFO("Back to previous window: %s", window->c_str());
+        windowsStack().popTo(*window);
```

Comment:
hmm popTo does not do switch here? it seems by name it should? or it is remove?

Reply:
Hmm it seems it is just remove specified window name -> so probably remove would be better ? or even pop with two implementations ? (with and without window).

Reply:
this is changed to the action which is actually happening with no needless wrapper. It's a stack so push and pop are valid names